File: sdk/ios/Sources/EngagelabCaptcha/DeviceInfo.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif
import CommonCrypto

/// Collects device-level signals for risk analysis.
public final class DeviceInfoCollector: @unchecked Sendable {

    public static let shared = DeviceInfoCollector()

    private init() {}

    /// Gather all device info into a string-keyed dictionary.
    public func collect() -> [String: String] {
        var info = [String: String]()

        #if canImport(UIKit)
        let device = UIDevice.current
        info["model"] = device.model
        info["system_name"] = device.systemName
        info["system_version"] = device.systemVersion
        info["device_name"] = device.name
        info["identifier_for_vendor"] = device.identifierForVendor?.uuidString ?? "unknown"

        if let screen = screenSize() {
            info["screen_width"] = "\(Int(screen.width))"
            info["screen_height"] = "\(Int(screen.height))"
            info["screen_scale"] = "\(Int(screen.scale))"
        }
        #else
        info["model"] = "unknown"
        info["system_name"] = "unknown"
        info["system_version"] = "unknown"
        #endif

        info["lang"] = Locale.preferredLanguages.first ?? Locale.current.identifier
        info["timezone"] = TimeZone.current.identifier
        info["timezone_offset"] = "\(TimeZone.current.secondsFromGMT())"
        info["user_agent"] = userAgentString()
        info["fingerprint"] = generateFingerprint(from: info)

        return info
    }

    // MARK: - Screen

    #if canImport(UIKit)
    private struct ScreenInfo {
        let width: CGFloat
        let height: CGFloat
        let scale: CGFloat
    }

    private func screenSize() -> ScreenInfo? {
        let bounds = UIScreen.main.bounds
        let scale = UIScreen.main.scale
        return ScreenInfo(width: bounds.width, height: bounds.height, scale: scale)
    }
    #endif

    // MARK: - User Agent

    /// Build a simple user-agent string describing this SDK and the device.
    private func userAgentString() -> String {
        let sdkVersion = "1.0.0"
        #if canImport(UIKit)
        let device = UIDevice.current
        return "EngagelabCaptcha-iOS/\(sdkVersion) (\(device.model); \(device.systemName) \(device.systemVersion))"
        #else
        return "EngagelabCaptcha-iOS/\(sdkVersion)"
        #endif
    }

    // MARK: - Fingerprint

    /// Generate a simple SHA-256 fingerprint hash from the collected info values.
    public func generateFingerprint(from info: [String: String]) -> String {
        let sortedKeys = info.keys.sorted()
        let combined = sortedKeys.map { "\($0)=\(info[$0] ?? "")" }.joined(separator: "&")
        return sha256(combined)
    }

    private func sha256(_ input: String) -> String {
        guard let data = input.data(using: .utf8) else { return "" }
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes { buffer in
            _ = CC_SHA256(buffer.baseAddress, CC_LONG(data.count), &hash)
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }
}

File: sdk/android/captcha/src/main/java/com/engagelab/captcha/DeviceInfo.kt
package com.engagelab.captcha

import android.content.Context
import android.content.res.Resources
import android.os.Build
import android.util.DisplayMetrics
import java.security.MessageDigest
import java.util.Locale
import java.util.TimeZone

/**
 * Collects device-level signals for risk analysis.
 */
object DeviceInfoCollector {

    /**
     * Gather all device info into a string-keyed map.
     *
     * @param context Android context, used to read display metrics and configuration.
     */
    fun collect(context: Context): Map<String, String> {
        val info = mutableMapOf<String, String>()

        // Hardware
        info["model"] = Build.MODEL
        info["manufacturer"] = Build.MANUFACTURER
        info["brand"] = Build.BRAND
        info["device"] = Build.DEVICE
        info["hardware"] = Build.HARDWARE
        info["product"] = Build.PRODUCT

        // OS
        info["os_version"] = Build.VERSION.RELEASE
        info["sdk_int"] = Build.VERSION.SDK_INT.toString()

        // Screen
        val dm: DisplayMetrics = Resources.getSystem().displayMetrics
        info["screen_width"] = dm.widthPixels.toString()
        info["screen_height"] = dm.heightPixels.toString()
        info["screen_density"] = dm.density.toString()
        info["screen_dpi"] = dm.densityDpi.toString()

        // Locale / Language
        val locale = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            context.resources.configuration.locales[0]
        } else {
            @Suppress("DEPRECATION")
            context.resources.configuration.locale
        }
        info["lang"] = locale.language
        info["lang_tag"] = locale.toLanguageTag()
        info["country"] = locale.country

        // Timezone
        val tz = TimeZone.getDefault()
        info["timezone"] = tz.id
        info["timezone_offset"] = (tz.rawOffset / 1000).toString()

        // User agent
        info["user_agent"] = buildUserAgent()

        // Emulator detection
        info["is_emulator"] = if (isEmulator()) "true" else "false"

        // Fingerprint
        info["fingerprint"] = generateFingerprint(info)

        return info
    }

    /**
     * Build a user-agent string describing this SDK and the device.
     */
    private fun buildUserAgent(): String {
        val sdkVersion = "1.0.0"
        return "EngagelabCaptcha-Android/$sdkVersion (${Build.MODEL}; Android ${Build.VERSION.RELEASE})"
    }

    /**
     * Detect common emulator signatures.
     */
    fun isEmulator(): Boolean {
        return (Build.FINGERPRINT.startsWith("generic")
                || Build.FINGERPRINT.startsWith("unknown")
                || Build.MODEL.contains("google_sdk")
                || Build.MODEL.contains("Emulator")
                || Build.MODEL.contains("Android SDK built for x86")
                || Build.MANUFACTURER.contains("Genymotion")
                || (Build.BRAND.startsWith("generic") && Build.DEVICE.startsWith("generic"))
                || Build.PRODUCT == "google_sdk"
                || Build.PRODUCT == "sdk"
                || Build.PRODUCT == "sdk_x86"
                || Build.PRODUCT == "sdk_gphone_x86"
                || Build.PRODUCT == "sdk_gphone64_arm64"
                || Build.PRODUCT == "vbox86p"
                || Build.HARDWARE.contains("goldfish")
                || Build.HARDWARE.contains("ranchu"))
    }

    /**
     * Generate a SHA-256 fingerprint hash from the collected info values.
     */
    fun generateFingerprint(info: Map<String, String>): String {
        val sortedKeys = info.keys.sorted()
        val combined = sortedKeys.joinToString("&") { key -> "$key=${info[key] ?: ""}" }
        return sha256(combined)
    }

    private fun sha256(input: String): String {
        val bytes = MessageDigest.getInstance("SHA-256").digest(input.toByteArray(Charsets.UTF_8))
        return bytes.joinToString("") { "%02x".format(it) }
    }
}
